[GH-24564] tailsamplingprocessor added invert_match rule for numeric …

LaserPhaser opened the pull request:
**Description:** <Describe what has changed.>
Added invert_match rule for numeric matcher, to support exclusion decision

**Link to tracking Issue:** 
[GH-24563]

**Testing:**
Added tests

**Documentation:**
Updated Readme

LaserPhaser committed:
[GH-24564] tailsamplingprocessor added invert_match rule for numeric matchers

jpkrohling commented on processor/tailsamplingprocessor/config.go:
```diff
@@ -139,6 +139,8 @@ type NumericAttributeCfg struct {
 	MinValue int64 `mapstructure:"min_value"`
 	// MaxValue is the maximum value of the attribute to be considered a match.
 	MaxValue int64 `mapstructure:"max_value"`
+	// If invertMatch is true it means that we will do the vice versa decision.
```
I would prefer something closer to what the String's InvertMatch states:

```
	// InvertMatch indicates that values must not match against attribute values.
	// If InvertMatch is true and Values is equal to '123', all other values will be sampled except '123'.
	// Also, if the specified Key does not match any resource or span attributes, data will be sampled.
```

jpkrohling replied:
It might be worth clarifying that the values are exclusive: a config with 10 as lower and 100 as higher limits means that values at 10 and 100 will NOT be sampled.

LaserPhaser replied:
I totally agree that explanation here should be the same as for string. Fixed. 
Could you please also review my proposal on [GH discussions](https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/24571)

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/numeric_tag_filter.go:
```diff
@@ -41,9 +43,13 @@ func (naf *numericAttributeFilter) Evaluate(_ context.Context, _ pcommon.TraceID
 			}
 		}
-		return false
+		if !naf.invertMatch {
```
Can't you just return the value of the invertMatch attribute?

```suggestion
		return naf.invertMatch
```

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/numeric_tag_filter_test.go:
```diff
@@ -65,6 +65,56 @@ func TestNumericTagFilter(t *testing.T) {
+			Decision: Sampled,
+		},
+		{
+			Desc:     "span attribute with lower limit",
```
```suggestion
			Desc:     "span attribute at the lower limit",
```

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/numeric_tag_filter_test.go:
```diff
@@ -65,6 +65,56 @@ func TestNumericTagFilter(t *testing.T) {
+			Decision: NotSampled,
+		},
+		{
+			Desc:     "span attribute with upper limit",
```
```suggestion
			Desc:     "span attribute at the upper limit",
```

LaserPhaser committed:
[GH-24564] review fixes

Added clear explanation to invert_match config string.
Fixed spelling.
Simpified logic of return sample decision.
